File: citadel/CitGame.py
import CitPlayer as CP
import CitBoard as CB
import CitOutAbstract
import CitInAbstract
import numbers
# ------------------------------------
# CLASS CitGame                      |
#                                    |
# Contains game logic and maintains  |
# overall game state                 |
# ------------------------------------
class CitGame( object ):

  def __init__( self ):
    self.P1 = 1
    self.P2 = 2
    self.__board = CB.CitBoard()
    self.__player = { self.P1 : CP.CitPlayer(), self.P2 : CP.CitPlayer() }
    self.__output = None
    self.__input = { self.P1 : None, self.P2 : None }

  def run( self ):
    winner = self.hasWinner()
    while winner == 0:
      self.showState()
      self.move( self.getMove( self.P1 ), self.getMove( self.P2 ) )
      winner = self.hasWinner()
    self.showState()
    self.showResult()
  
  # Set the desired move input module for the specified player ( P1 | P2 )
  def setInput( self, player, input_ ):
    if isinstance( input_, CitInAbstract.CitInAbstract ):
      try:
        self.__input[ player ] = input_
      except KeyError:
        raise Exception( 'CitGame.setInput(): Not a valid player, use P1 - 1 or P2 - 2' )
    else:
      raise Exception( 'CitGame.setInput(): Not a valid input object' )
      

  def setOutput( self, output ):
    if isinstance( output, CitOutAbstract.CitOutAbstract ):
      self.__output = output
    else:
      raise Exception( 'CitGame.setOutput(): Not a valid output object' )

  # Play out a turn of the game
  # p1m : Player ones move
  # p2m : Player twos move
  def move( self, p1m, p2m ):
    if ( isinstance( p1m, numbers.Number ) and
         isinstance( p2m, numbers.Number ) ):
      # AddMove() must not be called until both players moves have been recieved
      self.__player[ 1 ].addMove( p1m )
      self.__player[ 2 ].addMove( p2m )
      if p1m > p2m:
        self.__board.moveRight()
      elif p2m > p1m:
        self.__board.moveLeft()
    else:
      raise Exception( 'CitGame.move(): Not a number' );

  # Check if there == a winner
  # returns 0 : game not finished
  #         1 : player one
  #         2 : player two
  #         3 : draw
  def hasWinner( self ):
    if self.__board.getPosition() == CB.CitBoard.MAX:
      return 1
    if self.__board.getPosition() == CB.CitBoard.MIN:
      return 2
    if self.__player[ 1 ].hasLost():
      if self.__player[ 2 ].canWin( self.__board.getPosition() ):
        while self.__board.getPosition() > CB.CitBoard.MIN:
          self.__player[ 2 ].addMove( 1 )
          self.__board.moveLeft()
        return 2
      else:
        return 3
    if self.__player[ 2 ].hasLost():
      if self.__player[ 1 ].canWin( CB.CitBoard.MAX - self.__board.getPosition() ):
        while self.__board.getPosition() < CB.CitBoard.MAX:
          self.__player[ 1 ].addMove( 1 )
          self.__board.moveRight()
        return 1
      else:
        return 3
    return 0

  # Get a move for the specified player using the players input module
  def getMove( self, player ):
    # Calculate opponent from player value, 
    opponent = 3 # if player is invalid opponent will stay invalid as 3
    if player == self.P1:
      opponent = self.P2
    elif player == self.P2:
      opponent == self.P1
    # Try and call the players input module
    try:
      if self.__input[ player ] is not None:
        # TODO add name attribute and associated methods to CitPlayer class
        self.__input[ player ].getMove( str( player ),
                                        self.__player[ player ].getPoints(),
                                        self.__player[ opponent ].getLastMove() )
      else:
        # TODO is there a way to make this an uncatchable fatal error?
        raise Exception( 'CitGame.getMove(): Fatal error - No input module provided for ' + str( player ) )
    except KeyError:
      raise Exception( 'CitGame.getMove(): Not a valid player, use P1 - 1 or P2 - 2' )

  def showState( self ):
    if self.__output is not None:
      self.__output.showState( self.__board.getPosition(),
                               self.__player[ 1 ].getPoints(), 
                               self.__player[ 2 ].getPoints(),
                               self.__player[ 1 ].getLastMove(),
                               self.__player[ 2 ].getLastMove() )

  def showResult( self ):
    if self.__output is not None:
      self.__output.showResult( self.hasWinner() )
# ------------------------------------

File: javascript/citadel/CitGame.js
/******************************************************************************
 * Contains game logic and maintains overall game                            */
function CitGame( state ) {
  // Turn the correct PES buttons on
  var pinStat = { "UP": true, "DOWN": true, "LEFT": false, "RIGHT": false,
                  "SELECT": false, "START": true, "A": true, "B": false };
  $.ajax( {
    type: "POST",
    url: "http://10.0.0.1:8080/pins",
    contentType: "application/json",
    dataType: "json",
    data: pinStat,
  }).done( function() {
    console.log( 'PES button setup complete' );
  }).fail( function() {
    console.log( 'PES button setup failed' );
  });
  this.winner = 0;
  this.help   = 1;
  this.move   = 1;
  this.delay  = 0;
  this.board  = new CitBoard();
  this.bot    = new CitBot();
  this.player = { 1 : new CitPlayer( 'Player 1' ),
                  2 : new CitPlayer( 'Player 2' ) };
  this.state = state;
}
CitGame.prototype.ajax = function() {
  var self = this;
  $.ajax( {
    type: "GET",
    url: "http://10.0.0.1:8080/pressed",
    dataType: "json",
  }).done( function( data, textStatus, jqXHR  ) {
    self.handleInput( data.button );
    // After the request has returned call again
    self.ajax();
  }).fail( function( jqXHR, textStatus, errorThrown ) {
    console.log( 'Button request failed: ' + textStatus );
    // After the request has returned call again
    self.ajax();
  });  
};
CitGame.prototype.handleInput = function( button ) {
  if ( !this.winner ) {
    switch ( button ) {
      case 'UP':
        // Increase the propsed move by one
        if ( !( CitUpdateInstructionsVisible( this.state ) ) &&
              ( CitUpdateBoardVisible( this.state ) ) ) {
          if ( this.move < this.player[ 1 ].getPoints() ) {
            ++this.move;
            CitUpdateBid( this.state, this.move );
          }
        }
        break;
      case 'DOWN':
        // Decrease the proposed move by one
        if ( !( CitUpdateInstructionsVisible( this.state ) ) &&
              ( CitUpdateBoardVisible( this.state ) ) ) {
          if ( this.move > 1 ) {
            --this.move;
            CitUpdateBid( this.state, this.move );
          }
        }
        break;
      case 'A':
        if ( !( CitUpdateInstructionsVisible( this.state ) ) &&
              ( CitUpdateBoardVisible( this.state ) ) ) {
          CitUpdateToggleMoveBoard( this.state, this.move );
          // Play the current amount of points and get a move from the bot
          this.player[ 2 ].addMove( this.bot.getMove( this.player[ 2 ].getPoints(),
                                                      this.player[ 1 ].getLastMove() ) );
          // Add the human players move second so the bot doesn't know what the move is
          this.player[ 1 ].addMove( this.move );
          // Perform the move
          if ( this.player[ 1 ].getLastMove() > this.player[ 2 ].getLastMove() ) {
            this.board.moveRight();
          } else if ( this.player[ 1 ].getLastMove() < this.player[ 2 ].getLastMove() ) {
            this.board.moveLeft();
          }
          // Update the view
          CitUpdatePos( this.state, this.board.getPosition() );
          CitUpdatePoints( this.state, this.player[ 1 ].getPoints(), this.player[ 2 ].getPoints() );
          if ( this.move > this.player[ 1 ].getPoints() ) {
            this.move = this.player[ 1 ].getPoints();
          }
          // Set a 3 second delay that will ignore any button presses except help (START)
          this.delay = 1;
          var self = this;
          setTimeout( function() { 
            self.delay = 0;
            if ( !self.winner ) {
              CitUpdateToggleMoveBoard( self.state, self.move );
            } 
          }, 3000 );
        }
        break;
      case 'START':
        // Display the help
        CitUpdateInstructions( this.state );
        break;
      default: // Do nothing
        break;
    }
    // Check if the game is over 
    this.winner = this.hasWinner();
    var msg;
    if ( this.winner ) {
      switch ( this.winner ) {
        case 1:
          for ( var i = this.board.getPosition() ; i < CitBoard.MAX ; ++i ) {
            this.board.moveRight();
          }
          msg = 'Congratulations,\nyou won!';
          break;
        case 2:
          for ( var i = this.board.getPosition() ; i > 0 ; --i ) {
            this.board.moveLeft();
          }
          msg = 'Oh no, you lost!';
          break;
        case 3:
          msg = 'It\'s a draw! Boring.';
          break;
        default:
          msg = 'Game Over';
          break;
      }
      CitUpdateMsg( this.state, msg );
      CitUpdatePos( this.state, this.board.getPosition() );
      // Return to main menu or if single game refresh the game when finished
      setTimeout( function() { 
        var url = window.location.href;
        url = url.substring( 0, url.lastIndexOf( "/" ) + 1 ) + "../index.html";
        window.location.replace( url );
      }, 5000 );
    }
  }
};
/* Check if there is a winner
 * returns - 0:not finished, 1:player 1, 2:player 2, 3:draw */
CitGame.prototype.hasWinner = function() {
  if ( this.board.getPosition() == this.board.MAX ) return 1;
  if ( this.board.getPosition() == this.board.MIN ) return 2;
  if ( this.player[ 1 ].hasLost() ) {
    if ( this.player[ 2 ].canWin( this.board.getPosition() ) ) {
      while ( this.board.getPosition() > this.board.MIN ) {
        this.player[ 2 ].addMove( 1 );
        this.board.moveLeft();
      }
      return 2;
    } else {
      return 3;
    }
  }
  if ( this.player[ 2 ].hasLost() ) {
    if ( this.player[ 1 ].canWin( this.board.getPosition() ) ) {
      while ( this.board.getPosition() < this.board.MAX ) {
        this.player[ 1 ].addMove( 1 );
        this.board.moveRight();
      }
      return 1;
    } else {
      return 3;
    }
  }
  return 0;
};
CitGame.prototype.run = function() {
  var self = this;
  CitUpdateToggleMoveBoard( this.state, this.move );
  // Start ajax calls to RESTful api
  this.ajax();
  // Allows for keyboard to simulate pes input
  $( 'html' ).keydown( function ( e ) {
    var keyMap = { 37 : 'LEFT', 39 : 'RIGHT', 38 : 'UP', 40 : 'DOWN',
                   13 : 'START', 83 : 'SELECT', 65 : 'A', 66 : 'B' };
    if ( e.which in keyMap ) {
      self.handleInput( keyMap[ e.which ] );
    }
  });
};
/*****************************************************************************/
